### PR TITLE
Week12

### PR DESCRIPTION
**Pull request by viniciuspietscher**
- Add and serve static files;
- Update jobs controller;
- Set up delete job for front end;

### CODE REVIEW

File: controllers/jobs.js [hunk 1]
```diff
@@ -3,11 +3,20 @@ const { StatusCodes } = require("http-status-codes")
...
-  res.send("get all jobs")
+  const jobs = await Job.find({ createdBy: req.user.userId }).sort("createdAt")
+  res.status(StatusCodes.OK).json({ jobs, count: jobs.length })
 }
 
 const getJob = async (req, res) => {
-  res.send("get job")
+  const {
+    user: { userId },
+    params: { id: jobId },
```
Comment (jdevries3133) [inline controllers/jobs.js:13]: Wow I have never seen this nested destructing syntax in my life, very cool

File: controllers/jobs.js [hunk 2]
```diff
@@ -17,11 +26,35 @@ const createJob = async (req, res) => {
 }
 
 const updateJob = async (req, res) => {
-  res.send("update job")
+  const {
+    user: { userId },
+    body: { company, position },
```
Comment (jdevries3133) [inline controllers/jobs.js:31]: mind = blown

File: controllers/jobs.js [hunk 3]
```diff
@@ -17,11 +26,35 @@ const createJob = async (req, res) => {
...
+    params: { id: jobId },
+  } = req
+  if (company === "" || position === "") {
```
Comment (jdevries3133) [inline controllers/jobs.js]: Do you know a way to write this expression shorter? If so, why did you use this instead? If not, try to find one!!

File: middleware/error-handler.js [hunk 4]
```diff
@@ -1,10 +1,32 @@
...
-  if (err instanceof CustomAPIError) {
-    return res.status(err.statusCode).json({ msg: err.message })
+  let customError = {
+    // set default
+    statusCode: err.statusCode || StatusCodes.INTERNAL_SERVER_ERROR,
+    msg: err.message || "Something went wrong try again later",
   }
-  return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json({ err })
+
+  if (err.name === "ValidationError") {
```
Comment (jdevries3133) [inline middleware/error-handler.js:9]: I feel like for validation errors it would be an ideal to return some kind of JSON representation. Feel free to refactor towards that behavior if you wish

File: public/index.html [hunk 5]
```diff
@@ -0,0 +1,94 @@
+<!DOCTYPE html>
```
Comment (jdevries3133) [inline public/index.html:1]: TBR

File: public/jobs.js [hunk 6]
```diff
@@ -0,0 +1,349 @@
+async function buildJobsTable(jobsTable, jobsTableHeader, token, message) {
```
Comment (jdevries3133) [inline public/jobs.js:1]: Mmmm spaghetti (code) 🍝

Vinicius, any ideas on what we can do to improve this? Let's discuss & plan before jumping into programming.

**Review by jdevries3133 — COMMENTED** [review]
There's definitely room for improvement in the client-side script so I'm looking forward to discussing that! As always feel free to reach out with any questions. 

Also, I need to circle back to test locally since I'm on mobile right now, but everything looks solid.

File: controllers/jobs.js (hunk 2 above)
Comment (viniciuspietscher) [reply controllers/jobs.js:31]: right?! that was very cool indeed!

File: public/jobs.js (hunk 6 above)
Comment (viniciuspietscher) [reply public/jobs.js:1]: Could we create an object with the table elements and send as only one argument?

File: middleware/error-handler.js (hunk 4 above)
Comment (viniciuspietscher) [reply middleware/error-handler.js:9]: right now an object is returned with only a message and the status code:
image

should I return an example of how the request should look like?

File: controllers/jobs.js (hunk 3 above)
Comment (viniciuspietscher) [reply controllers/jobs.js]: I could do this, what do you think?

```suggestion
  if (!company || !position) {
```

Comment (jdevries3133) [reply controllers/jobs.js]: Hm better, but can you write it with just a single `!`?

File: middleware/error-handler.js (hunk 4 above)
Comment (jdevries3133) [reply middleware/error-handler.js:9]: I think the ideal thing to send back is something that matches up with the fields in the form so that server-side validation messages can be rendered on the client. For example:

```javascript
{
  username: [
    "must be at least 5 characters long",
    "cannot contain whitespace characters",
    "cannot contain special characters"
  ],
  password: [
    "cannot be blank",
    "must contain at least one special character",
    " must be at least 10 characters long"
  ]
}
```

So, on the server you would build up an object like this as you run your validation checks, then send it to the client so that the web app consuming your API can tell the end user *why* the form is invalid.

File: public/jobs.js (hunk 6 above)
Comment (jdevries3133) [reply public/jobs.js:1]: Well first, I'm not sure if this was clear, but I'm commenting on this whole script (the whole file), not this specific function. With that in mind, do you see any problems with the structure of this whole script? Especially comparing it to the much better server-side code, which is of course the focus of this project.

File: controllers/jobs.js (hunk 3 above)
Comment (viniciuspietscher) [reply controllers/jobs.js]: I didn't test, but would this work?

```suggestion
  if (!(company || position)) {
```

File: middleware/error-handler.js (hunk 4 above)
Comment (viniciuspietscher) [reply middleware/error-handler.js:9]: That's a great idea! Will work on implementing that!

File: controllers/jobs.js (hunk 3 above)
Comment (jdevries3133) [reply controllers/jobs.js]: What would happen for this input?

```javascript
const company = "Fancy Corp."
const position = ""

if (!(company || position)) {
  // the input is invalid, so we expect the code to go down this branch
  console.log('invalid input');
}

Comment (jdevries3133) [reply controllers/jobs.js]: You're close though!

Comment (viniciuspietscher) [reply controllers/jobs.js]: I just tested and it does not work.

It tries to update the job in the database and I get the custom error from mongodb.

I do not know how to write with a single `!`

File: public/jobs.js (hunk 6 above)
Comment (viniciuspietscher) [reply public/jobs.js:1]: now I see! it's just one single function with many conditionals. It makes it very difficult to read!

We could refactor and create distinct functions for each kind of event?

File: controllers/jobs.js (hunk 3 above)
Comment (jdevries3133) [reply controllers/jobs.js]: You just need this:

```javascript
if (!(company && position)) {...}
```

Comment (viniciuspietscher) [reply controllers/jobs.js]: that worked! Very cool!!! :grin:

Comment (jdevries3133) [reply controllers/jobs.js]: Now this is much more readable because it maps closely to describing the validation requirement in plain english: "if we do not have company and position."

However, you will have a problem because now our validation logic is more general; we don't validate that `company` and `position` are strings in the first place! Send this data to your API and watch it explode:

```json
{
  "company": {
    "oops": "you didn't expect an object here did you? Objects are truthy so it passes the check!"
  },
  "position": "Senior Bug Creator"
}
```

So, the most defensive implementation is probably

```javascript
if (!(company && postiion && typeof company === 'string' && typeof position === 'string')) { ... };
```

But, that's just one last thing to keep in mind.
